src/allmydata/crypto/rsa.py

```diff
@@ -0,0 +1,117 @@
+from cryptography.exceptions import InvalidSignature
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import hashes
+from cryptography.hazmat.primitives.asymmetric import rsa, padding
+from cryptography.hazmat.primitives.serialization import load_der_private_key, load_der_public_key, \
+    Encoding, PrivateFormat, PublicFormat, NoEncryption
+
+from allmydata.crypto import BadSignature
+
+
+class RsaMixin(object):
+
+    '''
+    This is the value that was used by `pycryptopp`, and we must continue to use it for
+    both backwards compatibility and interoperability.
+
+    The docs for `cryptography` suggest to use the constant defined at
+    `cryptography.hazmat.primitives.asymmetric.padding.PSS.MAX_LENGTH`, but this causes old
+    signatures to fail to validate.
+    '''
+    RSA_PSS_SALT_LENGTH = 32
+
+
+class PrivateKey(RsaMixin):
+
+    def __init__(self, priv_key):
+        self._priv_key = priv_key
+
+    @classmethod
+    def generate(cls, key_size):
+        priv_key = rsa.generate_private_key(
+            public_exponent=65537,  # serisously don't change this value
```

Comment:
Why is this the right value?

Reply:
I think "because the docs say so" and also presumably because this is what `pycryptopp` was using (I haven't confirmed that latter though).
https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa/#cryptography.hazmat.primitives.asymmetric.rsa.generate_private_key

Reply:
A link to https://cryptography.io/en/latest/hazmat/primitives/asymmetric/rsa/#cryptography.hazmat.primitives.asymmetric.rsa.generate_private_key and/or https://www.daemonology.net/blog/2009-06-11-cryptographic-right-answers.html in that comment would probably be a good thing.

Reply:
Yes, for sure .. I re-worked a bunch of this to be more "factory-function" and "helper-function"-like and will post a comparison tomorrow (i.e. removed the "wrapper" classes). I'm trying to convince myself what public-exponent pycryptopp actually uses, but haven't done that yet (sounds like "approximately everyone" uses this one though).

Reply:
I *think* that if I've traced the `pycryptopp` and `cryptopp` code properly, it'll use the `cryptopp` default for the public exponent for RSA, which is 17.

Reply:
This isn't quite proof but it might be interesting.  You can extract the public exponent from a generated key.

So I generated a key with KeyGenerator (on master) and extracted the public exponent:

```
>>> from allmydata.client import KeyGenerator                                                                                                                                                                                                                                 
>>> kg = KeyGenerator()                                                                                                                                                                                                                                                       
>>> open("/tmp/key.der", "w").write(kg.generate().result[0].serialize())                                                                                                                                                                                                      
>>>

[exarkun@baryon:~/Environments/tahoe]$ openssl rsa -inform DER -pubin -in /tmp/key.der -text -noout
RSA Public-Key: (2048 bit)
Modulus:
    00:ac:91:da:45:ce:9c:88:60:54:3a:35:de:c7:01:
    59:f3:f0:f6:39:59:f4:4d:61:88:31:ec:db:74:85:
    fb:ea:e8:a1:24:7b:f2:bb:7a:b3:2c:30:5b:c3:32:
    b1:e8:df:cc:1a:5a:1a:56:fc:04:c1:6d:dc:4a:14:
    9e:d5:8f:88:89:5a:9e:6e:a8:46:e3:90:0a:1d:21:
    1a:c8:6f:3d:00:7c:20:c4:d9:4d:bf:68:15:82:63:
    79:cb:61:b6:4d:f2:14:c4:49:de:ae:f1:91:a5:b7:
    6b:45:98:88:16:a0:57:7d:e7:d5:e6:29:ea:13:d0:
    2d:4c:ee:70:e0:aa:29:34:65:a2:d5:92:63:03:d1:
    44:fe:0c:55:52:b3:03:1b:34:39:90:46:f9:b2:9d:
    02:50:dd:6c:04:ca:ff:64:2f:18:d7:6a:12:0f:45:
    7a:fc:87:b3:b7:83:c5:bd:af:b3:e3:ee:55:d4:17:
    93:42:df:09:cd:a1:ea:ce:3f:dd:ab:59:e5:83:ec:
    93:26:b5:61:94:b2:4f:a2:0c:ad:1e:56:9f:19:00:
    d4:51:cf:dc:4d:a7:25:56:67:77:65:83:34:13:de:
    9c:67:10:2f:97:8b:db:32:81:14:64:2a:c7:e6:f6:
    c9:10:6f:2f:d3:40:b5:cd:ef:95:52:03:7e:76:2f:
    c6:29
Exponent: 17 (0x11)
```

So, probably, congratulations on your C++ tracing efforts. :)

Independently, I'm not sure it matters that we continue to use the same public exponent in keys that Tahoe-LAFS generates.  But I guess we have to decide now that we see that it used to be 17 and in this branch it's 65537.

Reply:
(see other thread, I've switched it back to 17 in this branch .. we can revisit this separately if need-be)